src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -0,0 +1,446 @@
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Analyzer.Utilities.PooledObjects;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.GlobalFlowStateAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.NetCore.Analyzers.InteropServices
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed partial class PlatformCompatabilityAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA1416";
+        private static readonly ImmutableArray<string> s_platformCheckMethodNames = ImmutableArray.Create("IsOSPlatformOrLater", "IsOSPlatformEarlierThan");
+        private static readonly ImmutableArray<string> s_osPlatformAttributes = ImmutableArray.Create(MinimumOSPlatformAttribute, ObsoletedInOSPlatformAttribute, RemovedInOSPlatformAttribute);
+
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.PlatformCompatabilityCheckTitle), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableAddedMessage = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.PlatformCompatibilityCheckAddedMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableObsoleteMessage = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.PlatformCompatabilityCheckObsoleteMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableRemovedMessage = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.PlatformCompatabilityCheckRemovedMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.PlatformCompatabilityCheckDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+
+        private const string MinimumOSPlatformAttribute = nameof(MinimumOSPlatformAttribute);
+        private const string ObsoletedInOSPlatformAttribute = nameof(ObsoletedInOSPlatformAttribute);
+        private const string RemovedInOSPlatformAttribute = nameof(RemovedInOSPlatformAttribute);
+        private const string TargetPlatformAttribute = nameof(TargetPlatformAttribute);
+        private const string Windows = nameof(Windows);
+        private const string OSX = nameof(OSX);
+        private const string macOS = nameof(macOS);
```

Comment:
I don't think we should hardcode OS platform names. Instead we should fetch the type symbol for well-known type name [System.Runtime.InteropServices.OSPlatform](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.osplatform?view=netcore-3.1) and then create a set by iterating over all it's static properties with the same type. This will ensure that analyzer keeps functioning in future when we add more platforms to `OSPlatform` struct.
```suggestion
```

Reply:
This is a special casing for OS name (not related to all, also Windows will be removed as we are not special casing it anymore), as this is provided as a string in the attribute constructor and comparisons made as a string I prefer to keep it as a hardcoded string, but will use nameof(OSPlatform.OSX); instead

Reply:
Why would we use fixed set of hard coded names and prevent the analyzer from working against all platforms and/or future platforms?

Reply:
Ah yes, I verified that you are only using these strings for normalizing platform names, not guarding the supported platforms. Probably add a comment on the same lines?